Unable to produce multiple synthetic beans of same type having different identifiers

### Describe the bug

I am producing multiple synthetic beans of same type without qualifier.
I am using the identifier to distinct them.

```java
@BuildStep
    @Record(ExecutionTime.STATIC_INIT)
    List<SyntheticBeanBuildItem> registerJdbcLockProviderInitializer(
            final ApplicationIndexBuildItem applicationIndexBuildItem,
            final DataSourceNameRecorder dataSourceNameRecorder) {
        final List<String> dataSourceNames = getDataSourcesNameFromJdbcSchedulerLocks(applicationIndexBuildItem);
        List<SyntheticBeanBuildItem> list = dataSourceNames.stream()
                .map(dataSourceName -> SyntheticBeanBuildItem.configure(DataSourceName.class)
                        .scope(Singleton.class)
                        .identifier(dataSourceName)
                        .supplier(dataSourceNameRecorder.dataSourceNameSupplier(dataSourceName))
                        .unremovable()
                        .done())
                .toList();
        return list;
    }
```

When running my integration test on **integration-tests** / **jdbc** it fails - while two synthetic beans are produced having **\<default\>** and **master** identifier - about already registered bean with same identifier.

```
[ERROR] Errors: 
[ERROR]   SchedulerLockResourceTest.testShedlockEndpoint » Runtime java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.SyntheticBeansProcessor#initStatic threw an exception: java.lang.IllegalStateException: A synthetic bean with identifier ouqbYaizsnxlxh2wrovrwNxvSIQ is already registered: SYNTHETIC bean [types=[io.quarkiverse.shedlock.providers.jdbc.runtime.DataSourceName, java.lang.Object], qualifiers=[@Default, @Any], target=n/a]
        at io.quarkus.arc.processor.BeanDeployment.addSyntheticBean(BeanDeployment.java:1472)
```



### Expected behavior

Two Synthetic beans should be produced and injected inside **JdbcLockProviderInitializer**

### Actual behavior

_No response_

### How to Reproduce?

1. git clone https://github.com/dcdh/quarkus-shedlock
2. mvn clean install

### Output of `uname -a` or `ver`

Linux 2a02-8428-dff8-c601-234b-8c10-a3c4-2308.rev.sfr.net 6.6.11-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 10 19:25:59 UTC 2024 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.9" 2023-10-17

### Quarkus version or git rev

3.8.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.1 (Red Hat 3.9.1-3)

### Additional information

_No response_

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/synthetic/SyntheticBeanBuildItemProxyTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/synthetic/SyntheticBeanBuildItemProxyTest.java
@@ -1,10 +1,11 @@
 package io.quarkus.arc.test.synthetic;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.lang.reflect.Method;
+import java.util.List;
 import java.util.function.Consumer;
 
 import jakarta.enterprise.context.ApplicationScoped;
@@ -14,6 +15,7 @@ import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
 import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem.ExtendedBeanConfigurator;
 import io.quarkus.builder.BuildChainBuilder;
@@ -46,20 +48,31 @@ public class SyntheticBeanBuildItemProxyTest {
                         // We need to use reflection due to some class loading problems
                         Object recorderProxy = bytecodeRecorder.getRecordingProxy(TestRecorder.class);
                         try {
-                            Method test = recorderProxy.getClass().getDeclaredMethod("test");
-                            Object proxy = test.invoke(recorderProxy);
-                            ExtendedBeanConfigurator configurator = SyntheticBeanBuildItem.configure(SynthBean.class)
+                            Method test = recorderProxy.getClass().getDeclaredMethod("test", String.class);
+
+                            Object proxy1 = test.invoke(recorderProxy, "ok");
+                            ExtendedBeanConfigurator configurator1 = SyntheticBeanBuildItem.configure(SynthBean.class)
                                     .scope(ApplicationScoped.class)
+                                    .identifier("ok")
                                     .unremovable();
                             // No creator
                             assertThrows(IllegalStateException.class,
-                                    () -> configurator.done());
+                                    () -> configurator1.done());
                             // Not a returned proxy
                             assertThrows(IllegalArgumentException.class,
-                                    () -> configurator.runtimeProxy(new SynthBean()));
-                            context.produce(configurator
-                                    .runtimeProxy(proxy)
+                                    () -> configurator1.runtimeProxy(new SynthBean()));
+                            context.produce(configurator1
+                                    .runtimeProxy(proxy1)
                                     .done());
+
+                            // Register a synthetic bean with same types and qualifiers but different identifier
+                            context.produce(SyntheticBeanBuildItem.configure(SynthBean.class)
+                                    .scope(ApplicationScoped.class)
+                                    .identifier("nok")
+                                    .unremovable()
+                                    .runtimeProxy(test.invoke(recorderProxy, "nok"))
+                                    .done());
+
                         } catch (Exception e) {
                             throw new RuntimeException(e);
                         }
@@ -73,9 +86,9 @@ public class SyntheticBeanBuildItemProxyTest {
     @Recorder
     public static class TestRecorder {
 
-        public SynthBean test() {
+        public SynthBean test(String val) {
             SynthBean bean = new SynthBean();
-            bean.setValue("ok");
+            bean.setValue(val);
             return bean;
         }
 
@@ -83,9 +96,12 @@ public class SyntheticBeanBuildItemProxyTest {
 
     @Test
     public void testBeans() {
-        SynthBean bean = Arc.container().instance(SynthBean.class).get();
-        assertNotNull(bean);
-        assertEquals("ok", bean.getValue());
+        List<InstanceHandle<SynthBean>> beans = Arc.container().listAll(SynthBean.class);
+        assertEquals(2, beans.size());
+        for (InstanceHandle<SynthBean> handle : beans) {
+            String val = handle.get().getValue();
+            assertTrue("ok".equals(val) || "nok".equals(val));
+        }
     }
 
     @Vetoed

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanConfiguratorBase.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanConfiguratorBase.java
@@ -72,6 +72,7 @@ public abstract class BeanConfiguratorBase<THIS extends BeanConfiguratorBase<THI
      */
     public THIS read(BeanConfiguratorBase<?, ?> base) {
         super.read(base);
+        identifier = base.identifier;
         types.clear();
         types.addAll(base.types);
         qualifiers.clear();

